Implemented Hopper pushing, pulling and picking up items logic

ColinHDev opened the pull request:
## Introduction
In pm4 hopper blocks were implemented to have an inventory. But the logic for pushing, pulling and picking up items was missing nonetheless.
This PR aims to add this logic to the hopper.
**This PR contains many comments because the hopper logic is very complex. To prevent future developers from getting confused about how certain things were done, most parts were commented to explain what was done and why.**

### Relevant issues
- Fixes https://github.com/pmmp/PocketMine-MP/blob/master/src/block/Hopper.php#L98-L102

### Sources
- https://minecraft.fandom.com/wiki/Hopper
  - https://static.wikia.nocookie.net/minecraft_gamepedia/images/3/38/Hopper_logic_flowchart.png/revision/latest?cb=20210518094503

## Changes
### API changes
- Added private function in block class `Hopper::push(HopperInventory) : bool` which handles pushing items from the hopper to a tile in the direction the hopper is facing.
- Added private function in block class `Hopper::pull(HopperInventory, Container) : bool` which handles pulling items by the hopper from a container above.
- Added private function in block class `Hopper::pickup(HopperInventory) : bool` which handles picking up items by the hopper.
- Added getter in tile class `Hopper::getTransferCooldown() : int` to get the transfer cooldown of that hopper.
- Added setter in tile class `Hopper::setTransferCooldown(int)` to set the transfer cooldown of that hopper.

### Behavioural changes
- Because hoppers need to be updated every tick, it could impact the server's performance if a large amount of hoppers is loaded.

### TODOs
- Pushing to specific blocks couldn't be implemented because these blocks aren't implemented at the moment. This includes the Composter, the Brewing Stand and the Jukebox (not working as in vanilla).
- Pulling from specific blocks couldn't be implemented because these blocks aren't implemented at the moment. This includes the Composter, the Brewing Stand and the Jukebox.
- Calling an event when a hopper moves an item between its own and another inventory.
- Calling an event when a hopper picks up an item.

## Backwards compatibility
- This PR is fully backwards compatible. Already placed hoppers should also work without any problems when updating to a version of pm4 which merged these changes.

## Tests
- Functionality tests
  - Every block a hopper can push items into was tested if it works as explained in the sources.
  https://www.youtube.com/watch?v=4gSyuViaPaU
  - Every block a hopper can pull items from was tested if it works as explained in the sources.
  https://www.youtube.com/watch?v=6NWvr6Kv88E
  - The collision box of the area where the hopper should pick up items was checked and works as explained in the sources.
  https://www.youtube.com/watch?v=hVEPiK9KWkA

- "Performance" tests
  - 128 hoppers pushing 27 * 64 dirt from one chest to another:
    https://timings.pmmp.io/?id=158627
    ![Screenshot (543)](https://user-images.githubusercontent.com/54852588/131256515-3611c594-08e1-45a1-8bd2-3ebbaf141c8a.png)

ColinHDev committed:
added getter and setter for the transfer cooldown

ColinHDev committed:
implement hopper logic for pushing, pulling and picking up items

ColinHDev committed:
added initiation of hopper block scheduling

ColinHDev committed:
added import of count function

ColinHDev committed:
added comment about a hopper's decision to pull or pick up items

ColinHDev committed:
hoppers now get their records removed if they push into a jukebox

ColinHDev committed:
the slot variable of the hopper inventory gets no longer overwritten while pushing into the furnace

dktapps commented:
I don't have the time to do an in-depth review right now, but one quick note: setting an `isNull` item into an inventory does the same thing as setting an air item. You don't need to check for it.

ColinHDev committed:
removed redundant Item::isNull() check

ColinHDev committed:
removed redundant import of ItemFactory

ColinHDev commented:
Oh, yeah, I see that now.
Thanks, I changed it

ColinHDev committed:
removed redundant check if the item is a shulkerbox while a hopper is pushing into one
# the shulkerbox inventory has already its own check if the item is a shulkerbox in the ShulkerboxInventory::canAddItem() function

sylvrs commented on src/block/Hopper.php:
```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+		$success = $this->push($tile);
+		// Hoppers that have a container above them, won't try to pick up items.
+		$origin = $this->position->getWorld()->getTile($this->position->getSide(Facing::UP));
+		if($origin instanceof Container){
```
This could be simplified to:
`$success |= $origin instanceof Container ? $this->pull($tile, $origin) : $this->pickup($tile);`

ColinHDev replied:
I especially didn't choose the ternary operator here because in the future, it can't be used without making everything crowded. 
Right now, it is easy to just use the Container interface as data type. But if Composters and Jukebox (Tiles that aren't Containers but still affected by hoppers) are fully implemented in the future, this if statement needs to change

sylvrs commented on src/block/Hopper.php:
```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+			$success |= $this->pickup($tile);
+		}
+		// The cooldown is only set back to 8 ticks if the hopper has done anything.
+		if((bool) $success){
```
This cast seems unnecessary given the fact that Hopper::push(), Hopper::pull(), and Hopper::pickup() all return booleans.

ColinHDev replied:
The bitwise operations cast $success at this point to an integer of 0 (false) or 1 (true).
I know that it is automatically cast to a boolean in the if statement but I made this to ensure that this definitely is a boolean and won't break in the future (idk.. if `declare(strict_types=1)` would stop auto casting in if statements in a future php version)

ColinHDev replied:
The bitwise operators were nice because they helped me implement the hopper logic (only reset cooldown if done anything) with just these few lines.
And I choose to cast it back to a boolean instead of comparing its value because it looked more clear to me.

sylvrs commented on src/block/Hopper.php:
```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+		}
+		// The cooldown is only set back to 8 ticks if the hopper has done anything.
+		if((bool) $success){
+			$tile->setTransferCooldown(8);
```
This should be moved to a constant. Hardcoding the value is not good practice.

ColinHDev replied:
Hm
I am unsure about a clear name for that constant.
TRANSFER_COOLDOWN doesn't suit in my opinion, because the variable that stores the current cooldown of the hopper already has just a slightly different name which could confuse.
But I don't think that MAX_TRANSFER_COOLDOWN is so perfect either.

sylvrs commented on src/block/Hopper.php:
```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+				}
+				// Hoppers pushing into empty hoppers set the empty hoppers transfer cooldown back to 8 ticks.
+				if(count($destination->getInventory()->getContents()) === 0){
+					$destination->setTransferCooldown(8);
```
See my above comment about using hardcoded values.

sylvrs commented on src/block/Hopper.php:
```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+					$itemInFurnace = $item->pop();
+				}
+
+				// TODO event on item inventory switch
```
I saw this in your main PR TODOs, but I believe that events should be included before this PR is merged.

ColinHDev replied:
Dylan already pointed out that this PR should only contain the logic itself and no events.
Also, the event for block picking up an item only exists since this commit https://github.com/pmmp/PocketMine-MP/commit/2139171a5525a89061cbadfa30f3f3f1243973b4 and we don't have an event for inventory interaction (InventoryTransactionEvent is not possible for this case, explained in https://github.com/pmmp/PocketMine-MP/issues/4405)

sylvrs reviewed (changes requested):
The PR overall seems good, but there are some nitpicks that I thought I'd point out 😄

dktapps commented on src/block/Hopper.php:
```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
+				}
+
+			}elseif($destination instanceof TileJukebox){
+				if(!$item instanceof Record){
```
Please surround this with braces like so:
```suggestion
				if(!($item instanceof Record)){
```
I know it's not functionally necessary, but it makes it easier to read.

dktapps reviewed (changes requested):
Not a depth review, but I have quite a few PRs to go over. Overall doesn't look bad, but I do find a lot of the comments unnecessary and there are a few other things I'm not too stoked about. I'll come back to this.

ColinHDev committed:
removed redundant comments and combined two similar if statements

ColinHDev committed:
TODO comments now match PMMP's coding style

ColinHDev committed:
added DEFAULT_TRANSFER_COOLDOWN constant

ColinHDev committed:
Merge remote-tracking branch 'origin/master' into hoppers

ColinHDev committed:
added braces around NOT operator

ColinHDev committed:
removed dependency on hopper tiles in logic functions

ColinHDev committed:
if a hopper couldn't push to a jukebox because the tile had no corresponding block, it won't count that as success

ColinHDev committed:
removed dependency on container tiles in pull logic function

ColinHDev committed:
fixed the provided parameter types for logic functions

ColinHDev committed:
added TODO about removing dependency on container tiles in pull logic if more blocks are added to that logic

ColinHDev committed:
used new inventory slot constants instead of hardcoded values

ColinHDev committed:
renamed smelting slot to input slot in comment

ColinHDev commented:
Changed everything except the bitwise operator for now.
If I removed it, the code would get a bit bigger from (for example)
```php
$success |= $this->pickup($tile);
```
to
```php
if($success){
    $this->pickup($tile);
}else{
    $success = $this->pickup($tile);
}
```

dktapps commented:
It can be written as `$success = $success && doSomething(...)` to stay purely within the bounds of boolean behaviour.

ColinHDev commented:
> It can be written as `$success = $success && doSomething(...)` to stay purely within the bounds of boolean behaviour.
It shouldn't at least according to both my knowledge and the wiki (I looked it up for that https://www.php.net/manual/en/language.operators.logical.php).
If pushing failed and `$success` would be false, the pull or pickup functions wouldn't be called

dktapps commented:
Oops. I meant `||`, not `&&`.

ColinHDev commented:
But then the functions wouldn't be called if the pushing succeeded and `$success` would be true

dktapps commented:
You're right. I guess the condition order would have to be flipped.

ColinHDev commented:
Yes
```php
$success = $this->pickup($inventory) || $success;
```
should do it.
It looks a little bit awkward in my opinion, that's why I wanted to use bitwise operators. But if it works and you think it suits better, I will change it

ColinHDev committed:
removed usage of bitwise operators

ColinHDev commented:
Changed everything, also bitwise operators. I would leave this PR open until https://github.com/pmmp/PocketMine-MP/pull/4413 is merged (or rejected) because its implementations of the BrewingStand and the improvements of the BrewingStandInventory would help m4e to also implement pushing and pulling logic for BrewingStands

ColinHDev commented:
I still don't have a good idea how to ditch dependency on Container files in the pickup logic.. It works for now, but how stated in the TODO in the future, not all blocks a hopper can pull from are containers
